### PR TITLE
CASSANDRA-11930

### PR DESCRIPTION
**Pull request by thobbs**
Test non-compound RangeTombstone ser during upgrades

Done for CASSANDRA-11930

### CODE REVIEW

File: upgrade_tests/cql_tests.py [hunk 1]
```diff
@@ -1602,6 +1603,86 @@ def cql3_insert_thrift_test(self):
             res = cursor.execute("SELECT * FROM test")
             assert rows_to_list(res) == [[2, 4, 8]], res
 
+    def cql3_non_compound_range_tombstones_test(self):
+        """
+        Checks that 3.0 serializes RangeTombstoneLists correctly
+        when communicating with 2.2 nodes.
+        @jira_ticket CASSANDRA-11930
+        """
+        cursor = self.prepare(start_rpc=True)
```
Comment (mambocab) [inline upgrade_tests/cql_tests.py]: This should be `session`, not `cursor` -- see the last paragraph [here](https://github.com/riptano/cassandra-dtest/blob/master/CONTRIBUTING.md#style).

File: upgrade_tests/cql_tests.py [hunk 2]
```diff
@@ -1602,6 +1603,86 @@ def cql3_insert_thrift_test(self):
...
+            name='cf',
+            column_type='Standard',
+            comparator_type='AsciiType',
+            key_validation_class='AsciiType',
+            default_validation_class='AsciiType',
+            column_metadata=column_defs)
+        client.system_add_column_family(cfdef)
+
+        # dumb schema agreement wait (we're using Thrift)
+        time.sleep(5.0)
```
Comment (mambocab) [inline upgrade_tests/cql_tests.py]: Anything wrong with `sesson.cluster.wait_for_schema_agreement()` here?

File: upgrade_tests/cql_tests.py [hunk 3]
```diff
@@ -1602,6 +1603,86 @@ def cql3_insert_thrift_test(self):
...
+                for i, column_name in enumerate(('a', 'b', 'c', 'd', 'e')):
+                    column_value = 'val{}'.format(i)
+                    client.batch_mutate(
+                        {key: {'cf': [Mutation(ColumnOrSuperColumn(column=Column(name=column_name, value=column_value, timestamp=100)))]}},
+                        ThriftConsistencyLevel.ALL)
+
+                # sanity check on the query
+                fetch_slice = SlicePredicate(slice_range=SliceRange('', '', False, 100))
+                row = client.get_slice(key, ColumnParent(column_family='cf'), fetch_slice, ThriftConsistencyLevel.ALL)
+                self.assertEqual(6, len(row))
```
Comment (mambocab) [inline upgrade_tests/cql_tests.py]: Could you please pass `row` as a 3rd argument here? and @c-kodman/@smccarthy788, could you make a note/file a ticket to change the assertion here once the `assertions.py` PR is merged?

File: upgrade_tests/cql_tests.py [hunk 4]
```diff
@@ -1602,6 +1603,86 @@ def cql3_insert_thrift_test(self):
...
+
+                # delete a slice of dynamic columns
+                slice_range = SliceRange('b', 'd', False, 100)
+                client.batch_mutate(
+                    {key: {'cf': [Mutation(deletion=Deletion(timestamp=101, predicate=SlicePredicate(slice_range=slice_range)))]}},
+                    ThriftConsistencyLevel.ALL)
+
+                # check remaining columns
+                row = client.get_slice(key, ColumnParent(column_family='cf'), fetch_slice, ThriftConsistencyLevel.ALL)
+                self.assertEqual(3, len(row))
```
Comment (mambocab) [inline upgrade_tests/cql_tests.py]: Same comment as above (ping @c-kodman @smccarthy788)

**Comment by mambocab** [comment]
Looks good, except for my nits, which are small. Trying to run locally now.

**Comment by thobbs** [comment]
These should be fixed, thanks.  I also forgot to push my minor changes to ``upgrade_base.py``, so I did that here.

File: upgrade_tests/upgrade_base.py [hunk 5]
```diff
@@ -95,11 +95,13 @@ def prepare(self, ordered=False, create_keyspace=True, use_cache=False,
...
 
-    def do_upgrade(self, session):
+    def do_upgrade(self, session, return_nodes=False):
         """
         Upgrades the first node in the cluster and returns a list of
         (is_upgraded, Session) tuples.  If `is_upgraded` is true, the
-        Session is connected to the upgraded node.
+        Session is connected to the upgraded node. If `return_nodes`
+        is True, a tuple of (is_upgraded, Session, Node) will be
+        returned instead.
```
Comment (mambocab) [inline upgrade_tests/upgrade_base.py]: :100: :100: :100:

File: upgrade_tests/upgrade_base.py [hunk 6]
```diff
@@ -135,12 +137,20 @@ def do_upgrade(self, session):
         sessions = []
```
Comment (mambocab) [inline upgrade_tests/upgrade_base.py]: Might be worth renaming this variable.

File: upgrade_tests/upgrade_base.py (hunk 5 above)
Comment (mambocab) [inline upgrade_tests/upgrade_base.py]: Filed a ticket internally to make some improvements to other code that calls this.

File: upgrade_tests/cql_tests.py [hunk 7]
```diff
@@ -1602,6 +1603,85 @@ def cql3_insert_thrift_test(self):
             res = cursor.execute("SELECT * FROM test")
             assert rows_to_list(res) == [[2, 4, 8]], res
 
+    def cql3_non_compound_range_tombstones_test(self):
```
Comment (mambocab) [inline upgrade_tests/cql_tests.py]: Ah, I forgot earlier: could you add a `@since` here, or otherwise version-gate in the test?

**Comment by mambocab** [comment]
Cool, thanks for addressing my nits, and maaaaaan am I glad to see that API change to `do_upgrade`, great thinking there.

I'll want to see a CassCI run on these changes -- maybe a run of some subset of the upgrade tests -- but I'm +1 on the code. Unless you've run these somewhere @thobbs, we'll figure it out in TE and let you know the result.

Just to confirm, we'll wait to merge until CASSANDRA-11930 is closed, yeah?

File: upgrade_tests/upgrade_base.py (hunk 5 above)
Comment (mambocab) [inline upgrade_tests/upgrade_base.py]: For future reference, [here's the ticket](https://datastax.jira.com/browse/CSTAR-575) (links to private JIRA)

**Comment by thobbs** [comment]
@mambocab Russ is working on a run of the upgrade tests in CI here: http://cassci.datastax.com/view/Dev/view/thobbs/job/thobbs-custom-11930-upgrade/.

> Just to confirm, we'll wait to merge until CASSANDRA-11930 is closed, yeah?

Correct.

File: upgrade_tests/upgrade_base.py (hunk 5 above)
Comment (thobbs) [inline upgrade_tests/upgrade_base.py]: Cool, I'll take a look at it.  I saw at least one spot that could benefit from it.

Comment (mambocab) [inline upgrade_tests/upgrade_base.py]: Not worth much time or churn right now, IMO, but yeah, have a look if you like!

**Pull request merged**